migrate deferred attachments when editing/deprecating a form

czue opened the pull request:
http://manage.dimagi.com/default.asp?234481

@millerdev the issue was that when you deprecate a form the ID changes, which was calling subsequent lookups to the blobdb to fail. this works around it by explicitly adding any couch attachments to `_deferred_blobs` during the deprecation process. open to other ideas about how to do this if you have them.

cc @emord

czue committed:
failing test for edit submissions with blobdb migration

czue committed:
actually test xml comparisons

czue committed:
migrate deferred attachments when editing/deprecating a form

emord commented on corehq/ex-submodules/couchforms/tests/test_edits.py:
```diff
@@ -12,7 +12,7 @@
 from couchforms.models import (
     UnfinishedSubmissionStub,
-)
+    XFormInstance)
```
is this the correct import style to use in hq?

czue replied:
hmm maybe not - i think my IDE just did this

emord commented:
Looks good to me. will let @millerdev take a pass

czue committed:
[ci skip] clean up imports

millerdev commented on corehq/form_processor/backends/couch/processor.py:
```diff
@@ -77,6 +77,21 @@ def apply_deprecation(cls, existing_xform, new_xform):
+                }
+            deprecated._deferred_blobs = {
+                name: _couch_attachments_to_deferred_dict(existing_xform, name, attachment_meta)
+                for name, attachment_meta in existing_xform._attachments.items()
```
I think we also want this to do the same for existing blobs from the blob db. To do that, this should be `existing_xform.blobs.items()`, but the fetch isn't going to work then. Also, rather than setting `deprecated._deferred_blobs` I think it would make sense to use a non-private API (the `_deferred_blobs` copy above is a bit of a hack). Something like this:
```py
for name, meta in existing_xform.blobs.items():
    deprecated.deferred_put_attachment(
        existing_form.fetch_attachment(name, stream=True),
        name=name,
        content_type=meta.content_type,
        content_length=meta.content_length,
    )
```

czue replied:
haha, i saw `deferred_put_attachment` in the doc string it also suggests that it shouldn't be used. will switch though.

i was wondering about the existing blob question, and specifically why it was working without any changes and only failing in this scenario but i didn't dig into that. fyi, I don't think your fetching code works because `existing_form` has already had its `_id` changed at this point. I can hack around that by temporarily changing the `_id` before calling it, but it's kind of messy. any other suggestions?

millerdev replied:
> i saw `deferred_put_attachment`. in the doc string it also suggests that it shouldn't be used.

haha, well the warning in that docstring definitely applies to `_deferred_blobs` as well

millerdev replied:
I think this blob-copying code should go above https://github.com/dimagi/commcare-hq/blob/c410f9bde7513bdd98d4f84e656c21a828692f8e/corehq/form_processor/parsers/form.py#L198

millerdev committed:
Copy persistent attachments on form deprecation

dimagimon commented on corehq/form_processor/parsers/form.py:
```diff
@@ -195,6 +195,15 @@ def apply_deprecation(existing_xform, new_xform, interface=None):
 
+    if existing_xform.persistent_blobs:
+        for name, meta in existing_xform.persistent_blobs.items():
+            with existing_form.fetch_attachment(name, stream=True) as content:
```
F821 undefined name 'existing_form'

czue commented:
looks good to me. :+1: on :white_check_mark:

emord commented:
```
======================================================================
ERROR: test_edit_submissions_simple (corehq.apps.commtrack.tests.test_rebuild.RebuildStockStateTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 232, in inner
    return helper(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 205, in __call__
    call_with_settings(fn_with_pre_and_post, run_config.settings, args, kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 220, in call_with_settings
    fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 201, in fn_with_pre_and_post
    self.fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/commtrack/tests/test_rebuild.py", line 123, in test_edit_submissions_simple
    form_id=form.form_id,
  File "/mnt/commcare-hq-ro/corehq/apps/hqcase/utils.py", line 54, in submit_case_blocks
    attachments=attachments,
  File "/mnt/commcare-hq-ro/corehq/apps/receiverwrapper/util.py", line 29, in submit_form_locally
    **kwargs
  File "/mnt/commcare-hq-ro/corehq/form_processor/submission_post.py", line 144, in run
    result = process_xform_xml(self.domain, self.instance, self.attachments)
  File "/mnt/commcare-hq-ro/corehq/form_processor/parsers/form.py", line 81, in process_xform_xml
    return _handle_id_conflict(instance, e.xform, domain)
  File "/mnt/commcare-hq-ro/corehq/form_processor/parsers/form.py", line 151, in _handle_id_conflict
    return _handle_duplicate(xform, instance)
  File "/mnt/commcare-hq-ro/corehq/form_processor/parsers/form.py", line 180, in _handle_duplicate
    existing_doc, new_doc = apply_deprecation(existing_doc, new_doc, interface)
  File "/mnt/commcare-hq-ro/corehq/form_processor/parsers/form.py", line 198, in apply_deprecation
    if existing_xform.persistent_blobs:
AttributeError: 'XFormInstanceSQL' object has no attribute 'persistent_blobs'
-------------------- >> begin captured stdout << ---------------------
<function test_edit_submissions_simple at 0x7f92051027d0> failed with the following settings:
settings.TESTS_SHOULD_USE_SQL_BACKEND = True
```

millerdev committed:
Fix XFormInstanceSQL interface

Make XFormInstanceSQL compatible with DeferredBlobMixin, which is a base class of XFormInstance.

millerdev commented on corehq/form_processor/parsers/form.py:
```diff
@@ -195,6 +195,15 @@ def apply_deprecation(existing_xform, new_xform, interface=None):
+                    name=name,
+                    content_type=meta.content_type,
+                    content_length=meta.content_length,
+                )
```
@snopoke: I'm assuming that attachments do not need to be copied when deprecating SQL forms. Does that sound correct to you? Context: https://github.com/dimagi/commcare-hq/pull/12750/commits/d95b941ce7682d86d70c1841e15a759c37348842

snopoke replied:
That's correct

millerdev committed:
Add deferred blobs to existing xform

...which will be getting a new for ID

benrudolph commented:
✅ , merging and deploying
